include/deal.II/sundials/kinsol.h

```diff
@@ -0,0 +1,664 @@
+     * a preconditioner (in a preconditioned iterative linear solver).
+     *
+     * The setup_jacobian() function is not called at every Newton iteration,
+     * but only as frequently as the solver determines that it is appropriate
+     * to perform the setup task. In this way, Jacobian-related data generated
+     * by setup_jacobian() is expected to be used over a number of Newton
+     * iterations.
+     *
+     * @param current_u Current value of u
+     * @param current_f Current value of F(u) or G(u)
+     *
+     * This function should return:
+     * - 0: Success
+     * - >0: Recoverable error (KINSOLReinit will be called if this happens, and
+     *       then last function will be attempted again
+     * - <0: Unrecoverable error the computation will be aborted and an assertion
+     *       will be thrown.
+     */
+    std::function<int(const VectorType &current_u,
+                      const VectorType &current_f)> setup_jacobian;
+
+    /**
+     * A function object that users may supply and that is intended to solve
+     * the Jacobian linear system. This function will be called by KINSOL
+     * (possibly several times) after setup_jacobian() has been called at least
+     * once. KINSOL tries to do its best to call setup_jacobian() the minimum
+     * amount of times. If convergence can be achieved without updating the
+     * Jacobian, then KINSOL does not call setup_jacobian() again. If, on the
+     * contrary, internal KINSOL convergence tests fail, then KINSOL calls
+     * again setup_jacobian() with updated vectors and coefficents so that
+     * successive calls to solve_jacobian_systems() lead to better convergence
+     * in the Newton process.
+     *
+     * If you do not specify a solve_jacobian_system() function, then a fixed
+     * point iteration is used instead of a Newton method. Notice that this may
+     * not converge, or may converge very slowly.
+     *
+     * The jacobian $J$ should be (an approximation of) the system Jacobian
+     * \f[
+     *   J = M - \gamma \frac{\partial f_I}{\partial y}
+     * \f]
+     * evaluated at `t`, `ycur`. `fcur` is $f_I(t,ycur)$.
+     *
+     * A call to this function should store in `dst` the result of $J^{-1}$
+     * applied to `src`, i.e., `J*dst = src`. It is the users responsability to
+     * set up proper solvers and preconditioners inside this function.
+     *
+     *
+     * Arguments to the function are
+     *
+     * @param[in] t  the current time
+     * @param[in] gamma  the current factor to use in the jacobian computation
+     * @param[in] ycur  is the current $y$ vector for the current KINSOL internal step
+     * @param[in] fcur  is the current value of the implicit right-hand side at ycur,
+     *        $f_I (t_n, ypred)$.
+     *
+     *
+     * This function should return:
+     * - 0: Success
+     * - >0: Recoverable error (KINSOLReinit will be called if this happens, and
+     *       then last function will be attempted again
```

Comment:
-> then THE last function

But I am not sure what "last function" refers to here. Would you mind clarifying?